### PR TITLE
🏗 Persist merge commit for reuse with later CI jobs

### PR DESCRIPTION
**Pull request by jridgewell**
This commit is purposefully forked several commits before (768808e58) the current master (d1105664). I'm going to rebase and merge to see what happens.

### CODE REVIEW

File: build-system/tasks/bundle-size/index.js
Comment (danielrozenberg) [inline build-system/tasks/bundle-size/index.js:169]: Super²nit: does it have to be `origin/master`? If we just set it to `master` would it work just the same?
Since this function is only supposed to be called during CI, basically the only difference is for us when we try to debug this locally - I can see myself wasting half a day not understanding why this doesn't work only to realize it's hardcoded for `origin` instead of `upstream`

**Comment by jridgewell** [comment]
First run forked from d110566, and CircleCI is "merging" it with c6b122b. Results:
- https://github.com/ampproject/amphtml/runs/2070096416
- https://app.circleci.com/pipelines/github/ampproject/amphtml/4036/workflows/0dea8d04-b2bc-414f-96eb-83d171b036b4/jobs/51919

Success so far!

**Comment by jridgewell** [comment]
I've now rebased on d876f3298, creating 0f8d04b. Let's see how this one goes.

File: build-system/tasks/bundle-size/index.js
Comment (jridgewell) [reply build-system/tasks/bundle-size/index.js:169]: Sure, I agree with this. I'm going to wait for the test runs to complete before changing.

**Comment by jridgewell** [comment]
The rebase failed, but for a surprising reason! After the build, but before bundle-size, a new commit was merged onto master. So we built from c2f098460a01f742f4e2c20865130448c83104c3 (https://app.circleci.com/pipelines/github/ampproject/amphtml/4038/workflows/dfff0a99-8d4a-400a-a39a-f5065b1fdb7b/jobs/51946/parallel-runs/0/steps/0-107), but diffed size against cfe8f2f5a04eea3cfcbf64be54e37af8c0e739fe (https://app.circleci.com/pipelines/github/ampproject/amphtml/4038/workflows/dfff0a99-8d4a-400a-a39a-f5065b1fdb7b/jobs/51966/parallel-runs/0/steps/0-107).

**Comment by rsimha** [comment]
> The rebase failed, but for a surprising reason! After the build, but before bundle-size, a new commit was merged onto master.

One way to avoid this is to report bundle sizes immediately after building in the same job (i.e., the old way). This probably means we'll need to merge the module and nomodule size reports on the server side instead of consolidating binaries during CI.

File: build-system/tasks/bundle-size/index.js
Comment (rsimha) [inline build-system/tasks/bundle-size/index.js:173]: I think the change in this PR should apply to both places that call this function (bundle size and visual diff). The intent behind this function is to provide a baseline for diffs, and I believe they should be identical. Perhaps reimplement the original function instead of adding a new one? WDYT?

**Comment by jridgewell** [comment]
The new persist logic worked! You can see we consistently used merge commit ba9a1678eef1a801555fe67938d07bcb0bfbef09, even though I purposefully merged https://github.com/ampproject/amphtml/pull/33062 after the build started (but before bundle-size started). This is proven by the changed diff lines:

```diff
*   ba9a1678e Justin Ridgewell - (HEAD -> pull/33165) Merge 0841e3d4929c06f6f0cbfa5938fe7594239e449d into 32a411de53fde79e349c6b636d8134acf070199d (4 minutes ago)
|\  
| * 0841e3d49 Justin Ridgewell - (origin/pull/33165) Fetch and store merge_commit at the beginning of the build (4 minutes ago)
- * | 32a411de5 Dima Voytenko - (origin/master, origin/HEAD, master) Multi-version: signal when extensions are known (#33162) (19 minutes ago)
+ * | 32a411de5 Dima Voytenko - Multi-version: signal when extensions are known (#33162) (27 minutes ago)
```

From that, we generated https://app.circleci.com/pipelines/github/ampproject/amphtml/4060/workflows/e50d7ea6-898d-455e-9c1e-ba6619708486/jobs/52305, and https://github.com/ampproject/amphtml/pull/33165/checks?check_run_id=2071240440.

**Comment by jridgewell** [comment]
I've now merged (using a merge commit) 6fc5bc61a, creating 7f4eaee8c.

**Comment by jridgewell** [comment]
And we have a successful https://app.circleci.com/pipelines/github/ampproject/amphtml/4064/workflows/1aa4cfb0-f652-4920-a78b-09e7dff2d126/jobs/52366/parallel-runs/0/steps/0-108 and https://github.com/ampproject/amphtml/pull/33165/checks?check_run_id=2071429754.

File: build-system/tasks/bundle-size/index.js
Comment (jridgewell) [reply build-system/tasks/bundle-size/index.js]: Outdated, since we're just persisting the merge commit now.

File: .circleci/config.yml
Comment (rsimha) [inline .circleci/config.yml:85]: Right now, we're doing a full checkout before saving the merge commit. This adds ~2 mins to the total CI round trip time (1.x mins for running the new job + a few seconds of waiting after the newly added stage).

image

It's also resulting in us hitting our account's concurrency limit.

Screen Shot 2021-03-09 at 7 03 38 PM

At the minimum, I think this step can be done more efficiently without doing a full checkout. Something like this works without any checking out of code:

```sh
$ git ls-remote https://github.com/ampproject/amphtml.git refs/pull/33165/merge
e36e2e05be16fe93ed41846827d14f4259d38cd6        refs/pull/33165/merge
```

A higher-level thought is that the problem of synchronizing baseline commits mainly applies to just those jobs that depend on downloading previously built artifacts, so I think we might be able to do away with this extra stage and just persist the commit at the time of building, and use that as the baseline while reporting statuses. Lemme do some homework on this to see if I can come up with a solution that works.

**Review by rsimha — CHANGES_REQUESTED** [review]
Thanks for investigating this and prototyping a solution! You've managed to discover the elusive crux of the problem: a commit sneaking into `master` between build stages of a CI build.

However, I think the solution can be made more efficient at the minimum, and potentially be done without adding a new CI stage. Posting an initial comment inline, will follow up after doing some homework.

**Comment by rsimha** [comment]
> And we have a successful [app.circleci.com/pipelines/github/ampproject/amphtml/4064/workflows/1aa4cfb0-f652-4920-a78b-09e7dff2d126/jobs/52366/parallel-runs/0/steps/0-108](https://app.circleci.com/pipelines/github/ampproject/amphtml/4064/workflows/1aa4cfb0-f652-4920-a78b-09e7dff2d126/jobs/52366/parallel-runs/0/steps/0-108) and [#33165 (checks)](https://github.com/ampproject/amphtml/pull/33165/checks?check_run_id=2071429754).

The "compare" link on this check [#33165 (checks)](https://github.com/ampproject/amphtml/pull/33165/checks?check_run_id=2071429754) shows code from other PRs, I believe it too needs to use the merge commit SHA instead of `gitCommitHash()`.

File: .circleci/config.yml
Comment (jridgewell) [reply .circleci/config.yml:85]: > `git ls-remote https://github.com/ampproject/amphtml.git refs/pull/33165/merge`

HOW DID YOU FIND THIS?!

> A higher-level thought is that the problem of synchronizing baseline commits mainly applies to just those jobs that depend on downloading previously built artifacts, so I think we might be able to do away with this extra stage and just persist the commit at the time of building

My real desire is to be able to persist the `.git` directory entirely, as in https://circleci.com/docs/2.0/caching/#source-caching. But we need to do https://github.com/ampproject/amphtml/issues/21691 first, so that the ridiculous 800mb of disk space is much smaller. Also, could attempt doing the full clone and commit log list in the Setup job, and let all other jobs do a shallow clone with the SHAs the Setup finds.

**Review by jridgewell — COMMENTED** [review]
> The "compare" link on this check #33165 (checks) shows code from other PRs, I believe it too needs to use the merge commit SHA instead of gitCommitHash().

I agree partially. The comparison is definitely wrong, but I think the [HEAD](https://github.com/ampproject/amphtml/commit/6b604381804ca118f463ce944da8c2f523f2946a) and [BASE](https://github.com/ampproject/amphtml/commit/c1212924d1f98052cea36b1a5606e3550a1a2cf5) are should be reported as they are. We should instead change the code changes link to [MERGE](https://github.com/ampproject/amphtml/commit/e36e2e05be16fe93ed41846827d14f4259d38cd6)?

File: .circleci/config.yml
Comment (rsimha) [reply .circleci/config.yml]: > My real desire is to be able to persist the `.git` directory entirely, as in [circleci.com/docs/2.0/caching/#source-caching](https://circleci.com/docs/2.0/caching/#source-caching).

Interesting.

> But we need to do #21691 first, so that the ridiculous 800mb of disk space is much smaller. 

We did use LFS for https://github.com/ampproject/amp-closure-compiler in the past, but quickly ran out of LFS quota for the `ampproject` enterprise GH account. If we try using LFS for `amphtml`, I'm sure we'll need to create a new payment plan.

image

> Also, could attempt doing the full clone and commit log list in the Setup job, and let all other jobs do a shallow clone with the SHAs the Setup finds.

Looks like the combination of docker and `git ls-remote` worked out better.

**Review by rsimha — COMMENTED** [review]
> I agree partially. The comparison is definitely wrong, but I think the [HEAD](https://github.com/ampproject/amphtml/commit/6b604381804ca118f463ce944da8c2f523f2946a) and [BASE](https://github.com/ampproject/amphtml/commit/c1212924d1f98052cea36b1a5606e3550a1a2cf5) are should be reported as they are. We should instead change the code changes link to [MERGE](https://github.com/ampproject/amphtml/commit/e36e2e05be16fe93ed41846827d14f4259d38cd6)?

Today, CI builds only report the base SHA to the bundle-size app (along with the actual sizes).

https://github.com/ampproject/amphtml/blob/c1212924d1f98052cea36b1a5606e3550a1a2cf5/build-system/tasks/bundle-size/index.js#L188-L191

The app infers the head SHA and [constructs](https://github.com/ampproject/amp-github-apps/blob/b1e6b29954974faf68a84afdf5e4771f59df7dcf/bundle-size/api.js#L118-L121) the comparison link. We'll likely need to change this API to properly fix the comparison link.

**Comment by jridgewell** [comment]
Added a `mergeSha` field to the bundle-size endpoint POST. We just need to update the API to use it.

File: .circleci/fetch_merge_commit.sh
Comment (rsimha) [inline .circleci/fetch_merge_commit.sh:31]: Now that this script no longer explicitly uses the `merge` command line arg, WDYT about splitting this into two scripts called `compute_merge_commit.sh` (`curl` this in the `Compute Merge Commit` step) and `fetch_merge_commit.sh` (this script, which essentially remains unchanged)?

(The first couple of early-exit sections can be copy-pasted to the extent necessary)

File: build-system/common/ci.js
Comment (rsimha) [inline build-system/common/ci.js:204]: The thumb-rule for every function in this file is that it should be usable on GH actions if ever required, and that it is based on well-known env vars from each platform.

https://github.com/ampproject/amphtml/blob/c1212924d1f98052cea36b1a5606e3550a1a2cf5/build-system/common/ci.js#L18-L24

For `CIRCLE_MERGE_SHA`, I wonder if we should just add a util function within `bundle-size/index.js`.

In addition, we'll need to use this as the baseline for visual diffs too.

https://github.com/ampproject/amphtml/blob/c1212924d1f98052cea36b1a5606e3550a1a2cf5/build-system/tasks/visual-diff/index.js#L136-L150

**Review by rsimha — COMMENTED** [review]
A few comments based on your latest changes.

**Comment by jridgewell** [comment]
Does GitHub linkify merge commits? bafa657

File: .circleci/config.yml
Comment (jridgewell) [reply .circleci/config.yml:40]: Renamed to `.CIRCLECI_WORKFLOW_MERGE_COMMIT`. We cannot use `.circleci/*` because the directory doesn't exist during fetch because we didn't clone the repo.

File: .circleci/config.yml
Comment (jridgewell) [reply .circleci/config.yml:77]: Done.

File: .circleci/config.yml
Comment (jridgewell) [reply .circleci/config.yml:81]: Done.

File: .circleci/fetch_merge_commit.sh
Comment (jridgewell) [reply .circleci/fetch_merge_commit.sh:31]: Because of the duplication, I actually think they should remain the same… I'mma leave this one to you.

File: build-system/common/ci.js
Comment (rsimha) [reply build-system/common/ci.js:204]: SG. Let's just rename this function to `circleciMergeSha` to clearly indicate that it's not a generic CI function.

File: .circleci/fetch_merge_commit.sh
Comment (rsimha) [reply .circleci/fetch_merge_commit.sh:31]: Fair deal. You've already gone above and beyond with this PR. 😃

**Review by rsimha — APPROVED** [review]
LGTM. Thanks for patiently dealing with all my review comments!

File: build-system/common/ci.js
Comment (rsimha) [inline build-system/common/ci.js:203]: Should this condition be `isPullRequestBuild() && isCircleciBuild()` to guarantee an empty value during push builds?

File: build-system/common/ci.js
Comment (jridgewell) [reply build-system/common/ci.js:204]: According to https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request, the `GITHUB_SHA` is actually the merge commit! And we're mixing up the CCI SHA (which is the HEAD SHA) with the GH SHA (which is the merge SHA) in `ciCommitSha` above.

I'll rename this for now, and leave it to you to cleanup the env vars.

**Comment by jridgewell** [comment]
Lol, deleting my branch borked master. I restored it until https://github.com/ampproject/amphtml/pull/33176 can be merged.

File: build-system/common/ci.js
Comment (rsimha) [reply build-system/common/ci.js]: `ciCommitSha` is only used during push builds, and according to [these docs](https://docs.github.com/en/actions/reference/environment-variables#default-environment-variables), `GITHUB_SHA` is the correct variable.

I will admit that `ci.js` has gotten confusing due to all the individual use cases, and could do with a separate cleanup.

Other cleanup is coming up in #33215.

File: .circleci/fetch_merge_commit.sh
Comment (rsimha) [reply .circleci/fetch_merge_commit.sh:31]: Cleanup coming up in #33215.

**Pull request merged**
